File: 22/2/main.go
package main

import (
	"fmt"
	"io/ioutil"
	"strings"
)

func main() {
	file, _ := ioutil.ReadFile("input.txt")
	lines := strings.Split(string(file), "\n")
	games := make([][]string, len(lines))
	for i, line := range lines {
		games[i] = strings.Split(line, " ")
	}
	fmt.Println(calculateScore(games))
	prepare(games)
	fmt.Println(calculateScore(games))
}

func prepare(games [][]string) {
	for i, game := range games {
		if len(game) != 2 {
			continue
		}
		switch game[1] {
		case "X":
			switch game[0] {
			case "A":
				games[i][1] = "Z"
			case "B":
				games[i][1] = "X"
			case "C":
				games[i][1] = "Y"
			}
		case "Y":
			switch game[0] {
			case "A":
				games[i][1] = "X"
			case "B":
				games[i][1] = "Y"
			case "C":
				games[i][1] = "Z"
			}
		case "Z":
			switch game[0] {
			case "A":
				games[i][1] = "Y"
			case "B":
				games[i][1] = "Z"
			case "C":
				games[i][1] = "X"
			}
		}
	}
}

func calculateScore(games [][]string) (res int) {
	for _, game := range games {
		if len(game) != 2 {
			continue
		}
		switch game[1] {
		case "X":
			res += 1
			switch game[0] {
			case "A":
				res += 3
			case "B":
			case "C":
				res += 6
			}
		case "Y":
			res += 2
			switch game[0] {
			case "A":
				res += 6
			case "B":
				res += 3
			case "C":
			}
		case "Z":
			res += 3
			switch game[0] {
			case "A":
			case "B":
				res += 6
			case "C":
				res += 3
			}
		}
	}

	return
}

File: 22/2/src/main.rs
fn main() {
    let converted_lines = include_str!("../input.txt")
        .lines()
        .map(|line| {
            line.split(" ")
                .map(|c| c.chars().collect::<Vec<char>>()[0])
                .collect::<Vec<char>>()
        })
        .map(|line| {
            line.iter()
                .map(|c| {
                    if c >= &'A' && c <= &'C' {
                        *c as u8 - 'A' as u8
                    } else {
                        *c as u8 - 'X' as u8
                    }
                })
                .map(Into::<i32>::into)
                .collect::<Vec<i32>>()
        })
        .collect::<Vec<Vec<i32>>>();
    println!(
        "{}",
        converted_lines
            .iter()
            .map(|line| {
                return (match (line[1] - line[0] + 3) % 3 {
                    0 => 3,
                    1 => 6,
                    2 => 0,
                    _ => 0,
                }) + line[1]
                    + 1;
            })
            .sum::<i32>()
    );
    println!(
        "{}",
        converted_lines
            .iter()
            .map(|line| {
                return (match line[1] {
                    0 => 0 + ((line[0]+2)%3),
                    1 => 3 + line[0],
                    2 => 6 + ((line[0]+1)%3),
                    _ => 0,
                })
                    + 1;
            })
            .sum::<i32>()
    )
}
